stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/DOM.java

```diff
@@ -72,44 +72,50 @@ public void disable(JsonRpcPeer peer, JSONObject params) {
   public JsonRpcResult getDocument(JsonRpcPeer peer, JSONObject params) {
     final GetDocumentResponse result = new GetDocumentResponse();
 
-    mDOMProvider.postAndWait(new Runnable() {
-       @Override
-       public void run() {
-         Object rootElement = mDOMProvider.getRootElement();
-         if (rootElement != null) {
-           result.root = createNodeForElement(rootElement);
-         }
-       }
-     });
+    result.root = mDOMProvider.callAndWait(new Callable<Node>() {
+      @Override
+      public Node call() {
+        Object rootElement = mDOMProvider.getRootElement();
+        return (rootElement == null) ? null : createNodeForElement(rootElement);
+      }
+    });
 
     return result;
   }
 
   @ChromeDevtoolsMethod
   public void highlightNode(JsonRpcPeer peer, JSONObject params) {
-    HighlightNodeRequest request = mObjectMapper.convertValue(params, HighlightNodeRequest.class);
+    final HighlightNodeRequest request = mObjectMapper.convertValue(params, HighlightNodeRequest.class);
     if (request.nodeId == null) {
       LogUtil.w("DOM.highlightNode was not given a nodeId; JS objectId is not supported");
-    } else {
-      final RGBAColor contentColor = request.highlightConfig.contentColor;
-      if (contentColor == null) {
-        LogUtil.w("DOM.highlightNode was not given a color to highlight with");
-      } else {
-        final Object element = mObjectIdMapper.getObjectForId(request.nodeId);
-
-        mDOMProvider.postAndWait(new Runnable() {
-          @Override
-          public void run() {
-            mDOMProvider.highlightElement(element, contentColor.getColor());
-          }
-        });
-      }
+      return;
+    }
+
+    final RGBAColor contentColor = request.highlightConfig.contentColor;
+    if (contentColor == null) {
+      LogUtil.w("DOM.highlightNode was not given a color to highlight with");
+      return;
     }
+
+    mDOMProvider.postAndWait(new Runnable() {
+      @Override
+      public void run() {
+        Object element = mObjectIdMapper.getObjectForId(request.nodeId);
+        if (element != null) {
+          mDOMProvider.highlightElement(element, contentColor.getColor());
```

Comment:
1) That's not the contract established by DOMProvider. It requires all calls into it be made on the thread that is stipulated by check/verifyAccess().

2) Yes but no (re: "doesn't this also punt to the main thread internally"). If you look at the code for AndroidDOMProvider.highlightElement(), it makes two calls after verifyAccess(). First it calls getHighlightingView(), which must run on the UI thread. Then it calls mHighlighter.clearHighlight() or setHighlightedView(). It doesn't marshal to the UI thread by itself. Those two methods on the highlighter then do their own post-ing but that's for debouncing purposes.